crates/bevy_ecs/src/query/fetch.rs

```diff
@@ -276,6 +277,9 @@ use variadics_please::all_tuples;
     note = "if `{Self}` is a component type, try using `&{Self}` or `&mut {Self}`"
 )]
 pub unsafe trait QueryData: WorldQuery {
+    /// True if this query is read-only and may not perform mutable access.
+    const IS_READ_ONLY: bool;
+
     /// The read-only variant of this [`QueryData`], which satisfies the [`ReadOnlyQueryData`] trait.
     type ReadOnly: ReadOnlyQueryData<State = <Self as WorldQuery>::State>;
```

Comment:
Is it possible to bound ReadOnly to require `IS_READ_ONLY == true`?

Reply:
I think it would require the unstable `generic_const_exprs` feature, since the bounds only take types and you can't currently refer to an associated constant in a type.  

I'm happy to update it if someone figures out how to do it!  We'd probably want to do the same for `QueryFilter::IS_ARCHETYPAL` / `ArchetypeFilter`.